### PR TITLE
Stop using ConcurrentBag in tests for complex types

### PR DESCRIPTION
**Pull request by halter73**
dotnet/corefx#23068

### CODE REVIEW

**Comment by benaadams** [comment]
`ConcurrentQueue` is always my favorite; also has a `.ToArray` method on it

File: test/shared/TestApplicationErrorLogger.cs [hunk 1]
```diff
@@ -29,7 +50,11 @@ public class TestApplicationErrorLogger : ILogger
 
         public IDisposable BeginScope<TState>(TState state)
         {
-            Scopes.Add(state);
+            lock (_lockObj)
+            {
+                _scopes.Add(state);
+            }
+
             return new Disposable(() => { });
```
Comment (davidfowl) [inline test/shared/TestApplicationErrorLogger.cs]: May as well remove it here.

**Comment by halter73** [comment]
@benaadams  Then we probably wouldn't need the new locks or ToArrays(). But now I'm afraid of the Concurrent types that aren't ConcurrentDictionary 😞

**Comment by benaadams** [comment]
ConcurrentBag is the only one that uses thread local storage; unordered and less contended (via thread local)

**Comment by halter73** [comment]
Attempt number 2.

**Pull request merged**